### PR TITLE
fully non mutating "munkres!" without copy

### PR DESCRIPTION
**Pull request by francescoalemanno**
Hi Gnimuc, with this tiny change your code can become fully non copying (for most array types),
since in this way munkres! actually does not mutate the cost matrix,
let me know what you think!

i'm keeping a copy of your code at https://github.com/francescoalemanno/SolveLAP.jl
since i plan on making further changes which would not adhere to your package,
In the files where i use your code i made an explicit NOTICE section to tell users that such code comes from your excellent package.

If my repository bothers you i will take it down immediately!
Thank you for Hungarian.jl, it is excellent

### CODE REVIEW

File: src/Munkres.jl [hunk 1]
```diff
@@ -67,15 +63,39 @@ function munkres!(costMat::AbstractMatrix{T}) where T <: Real
...
+	
     # "consider a row of the matrix A;
     #  subtract from each element in this row the smallest element of this row.
     #  do the same for each row."
-    costMat .-= minimum(costMat, dims=2)
-
+    for i in 1:rowNum
+        mw=typemax(T)
+        for j in 1:colNum
+            cost=costMat[i,j]+Δrow[i]+Δcol[j]
```
Comment (francescoalemanno) [inline src/Munkres.jl]: +Δrow[i]+Δcol[j]
this are not necessary, i added them for symmetry

File: src/Munkres.jl [hunk 2]
```diff
@@ -67,15 +63,39 @@ function munkres!(costMat::AbstractMatrix{T}) where T <: Real
...
+	
     # "then consider each column of the resulting matrix and subtract from each
     #  column its smallest entry."
     # Note that, this step should be omitted if the input matrix is not square.
-    rowNum == colNum && (costMat .-= minimum(costMat, dims=1);)
+    if rowNum == colNum
+        for j in 1:colNum
+            mw=typemax(T)
+            for i in 1:rowNum
+                cost=costMat[i,j]+Δrow[i]+Δcol[j]
```
Comment (francescoalemanno) [inline src/Munkres.jl]: +Δcol[j] 

this is also not necessary, i added it for symmetry

File: src/Munkres.jl [hunk 3]
```diff
@@ -67,15 +63,39 @@ function munkres!(costMat::AbstractMatrix{T}) where T <: Real
...
     Zs = spzeros(Int8, rowNum, colNum)
 
+    # for tracking changes per row/col of A
+    Δrow = zeros(rowNum)
+    Δcol = zeros(colNum)
+	
     # "consider a row of the matrix A;
     #  subtract from each element in this row the smallest element of this row.
     #  do the same for each row."
-    costMat .-= minimum(costMat, dims=2)
```
Comment (francescoalemanno) [inline src/Munkres.jl:73]: this kind of operation can be very expensive if the user wants to supply a custom matrix type with a complicated setindex! operation.
thus in my use case it is very advantageous to make the algorithm fully non mutating and non copying

File: src/Munkres.jl [hunk 4]
```diff
@@ -67,15 +63,39 @@ function munkres!(costMat::AbstractMatrix{T}) where T <: Real
...
+                mw=cost
+            end
+        end
+        Δrow[i] = -mw
+    end
+	
     # "then consider each column of the resulting matrix and subtract from each
     #  column its smallest entry."
     # Note that, this step should be omitted if the input matrix is not square.
-    rowNum == colNum && (costMat .-= minimum(costMat, dims=1);)
```
Comment (francescoalemanno) [inline src/Munkres.jl:78]: same risk as above

File: src/Munkres.jl [hunk 5]
```diff
@@ -26,7 +26,7 @@ julia> full(matching)
  0  0  2
 ```
 """
-munkres(costMat::AbstractMatrix{T}) where {T<:Real} = munkres!(copy(costMat))
```
Comment (francescoalemanno) [inline src/Munkres.jl:29]: this copy  is no longer necessary, 
i still left the copying operation in case the input matrix contains missings

**Review by francescoalemanno — COMMENTED** [review]
some explanations which are probably unnecessary

**Comment by Gnimuc** [comment]
Thanks for improving this! Fully non-mutating `munkres!` would be cool.

I did a benchmark against the master branch, here is the report:
https://gist.github.com/Gnimuc/aa8847bd3ff93296c08637ede8d8ad17 

With this PR, the memory allocation declined dramatically, but it looks like this PR also introduces some performance overhead. I guess we need to optimize the newly added loops. I'll revisit this later.

BTW, @Wikunia has a faster implementation for large arrays: https://opensourc.es/blog/hungarian-performance & https://github.com/opensourcesblog/hungarian

**Comment by francescoalemanno** [comment]
thank you for considering this :)

I think the core issue Is a type instability when the cost Matrix is not a
Float64, since the two delta vectors are Always allocated as F64

A partial solution could be changing zeros(...) to a zeros(T,...)

But this will break UInt tests, since UInt do not allow for Delta < 0

Anyway i due not test any of this so i could be wrong, today i will try to
fix It!

Thank you again!!

Il mer 4 mar 2020, 01:55 Gnimuc <notifications@github.com> ha scritto:

**Comment by francescoalemanno** [comment]
The benchmarks hopefully will be better now  :+1:

**Comment by Gnimuc** [comment]
Yep, you can find the new benchmark result here: https://gist.github.com/Gnimuc/aa8847bd3ff93296c08637ede8d8ad17#gistcomment-3200013

I also did a profile and the result looks just fine, I'm still not sure why there're regressions when the input array has a large size in the squared benchmark suite or when the first dimension of the input array is larger than the second dimension in the non-squared benchmark suite.

**Comment by Gnimuc** [comment]
The bottleneck should lie in step3, but this PR makes almost no change in the step3. There shouldn't be so many performance regressions. 🤔 I even suspect I made some mistakes when I was doing the benchmark.

**Comment by Wikunia** [comment]
@francescoalemanno have you tried adding `@inbounds` in front of your loops?
If there is still a performance decrease I am wondering whether it's reasonable to still have the `munkres!` function which changes the cost matrix but is slightly faster. 
Currently the `!` is misleading so you could deprecate it and only allow `munkres`? Probably no one relies on the changed matrix when using `munkres!` but you never know :smile:

**Comment by francescoalemanno** [comment]
@Wikunia

> @francescoalemanno have you tried adding `@inbounds` in front of your loops?

This should absolutely be done, it could be the culprit!

@Gnimuc 
> The bottleneck should lie in step3, but this PR makes almost no change in the step3. There shouldn't be so many performance regressions. thinking I even suspect I made some mistakes when I was doing the benchmark.

this is very surprising

EDIT ---------------

in the last commit i added @inbounds at loops and i added the super ugly goto in order to skip a whole cycle over a col|row  if the cost == 0,   

re-Edit:

@Gnimuc the reason for the increased cost on step3 was that i forgot to fix the preparatory steps to make use of the non zero Delta's, now it is fixed
i also removed the goto uglyness and stuff

File: src/Munkres.jl [hunk 6]
```diff
@@ -70,30 +70,34 @@ function munkres!(costMat::AbstractMatrix{T}) where T <: Real
...
     #  subtract from each element in this row the smallest element of this row.
     #  do the same for each row."
-    for i in 1:rowNum
+    @inbounds for i in 1:rowNum
         mw=typemax(T)
-        for j in 1:colNum
-            cost=costMat[i,j]+Δrow[i]+Δcol[j]
+        @inbounds for j in 1:colNum
+            cost=costMat[i,j]#+Δrow[i]+Δcol[j]
+	    iszero(cost) && @goto skip1
```
Comment (Wikunia) [inline src/Munkres.jl]: isn't this just a `break`?

File: src/Munkres.jl [hunk 7]
```diff
@@ -70,30 +70,34 @@ function munkres!(costMat::AbstractMatrix{T}) where T <: Real
...
     # Note that, this step should be omitted if the input matrix is not square.
     if rowNum == colNum
-        for j in 1:colNum
+        @inbounds for j in 1:colNum
             mw=typemax(T)
-            for i in 1:rowNum
-                cost=costMat[i,j]+Δrow[i]+Δcol[j]
+            @inbounds for i in 1:rowNum
+                cost=costMat[i,j]+Δrow[i]#+Δcol[j]
+		iszero(cost) && @goto skip2
```
Comment (Wikunia) [inline src/Munkres.jl]: looks like a `break` is enough to jump out

Comment (francescoalemanno) [reply src/Munkres.jl]: I wanted to Jump past the update to delta

Comment (Wikunia) [reply src/Munkres.jl]: Oh sorry missed that. Thanks for clarifying.

**Comment by francescoalemanno** [comment]
**i strongly believe it is fixed now!**
The benchmarks will confirm this i hope <3

<strike>edit1: already the total testing time on Travis decreased dramatically 14minutes to 5 minutes</strike>(Unrealiable)
edit2: some calculations i'm doing using Hungarian already got faster by a factor ~1.7 (on a DefaultArray, for simple Arrays it will be faster i think)

File: src/Munkres.jl [hunk 8]
```diff
@@ -85,11 +105,12 @@ function munkres!(costMat::AbstractMatrix{T}) where T <: Real
     row2colSTAR = Dict{Int,Int}()
     for ii in CartesianIndices(size(costMat))
         # "consider a zero Z of the matrix;"
-        if costMat[ii] == 0
```
Comment (francescoalemanno) [inline src/Munkres.jl:88]: This had to be fixed! cannot rely on costMat alone

File: src/Munkres.jl (hunk 7 above)
Comment (francescoalemanno) [reply src/Munkres.jl]: in the latest commit i removed such uglyness

**Comment by Gnimuc** [comment]
I tweaked the code a little bit, here is the benchmark report: https://gist.github.com/Gnimuc/aa8847bd3ff93296c08637ede8d8ad17#gistcomment-3201356

The result shows that we still get regressions for large `UInt16` arrays. I guess the reason is the `@simd` failed to vectorize the code in these cases. But I'm fine with these regressions. If users want peak performance, I would suggest using @Wikunia's implementation.

**Comment by Wikunia** [comment]
That looks much better! Oh don't send them over to my repo :smiley: not sure whether it's really faster...
Btw I love your easy way of doing benchmarks. Maybe I'll use it for my own repositories :wink:

**Comment by Gnimuc** [comment]
Another thought about improving the performance for large arrays is we could make use of [Julia's new multithreading feature introduced in 1.3](https://julialang.org/blog/2019/07/multithreading/) since each of these two loops can work in a parallel way.

**Comment by Wikunia** [comment]
Would be interesting to parallelize but I think that part is quite fast and the overhead is probably too high but I might be wrong.

**Pull request merged**